pkg/serviceprovider/common/common.go

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright (c) 2021 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	api "github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/serviceprovider"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/spi-shared/config"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Common is a unified provider implementation for any URL-based tokens and only supports
+// manual upload of token data. Matching is done only by URL of the provider, so it is possible to have
+// only one token for particular URL in the given namespace.
+type Common struct {
+	Configuration config.Configuration
+	lookup        serviceprovider.GenericLookup
+	httpClient    rest.HTTPClient
+	repoUrl       string
+}
+
+var Initializer = serviceprovider.Initializer{
+	Constructor: serviceprovider.ConstructorFunc(newCommon),
+}
+
+func newCommon(factory *serviceprovider.Factory, repoUrl string) (serviceprovider.ServiceProvider, error) {
+
+	cache := serviceprovider.NewMetadataCache(factory.KubernetesClient, &serviceprovider.NeverMetadataExpirationPolicy{})
+	return &Common{
+		Configuration: factory.Configuration,
+		lookup: serviceprovider.GenericLookup{
+			ServiceProviderType: api.ServiceProviderTypeCommon,
+			TokenFilter:         &tokenFilter{},
+			RepoHostParser: serviceprovider.RepoHostParserFunc(func(repoUrl string) (string, error) {
+				schemeIndex := strings.Index(repoUrl, "://")
```

Comment:
Since this is just copied from Quay, I think it would make sense to actually extract this to a separate function, say `RepoHostFromSchemelessUrl` and use it here and in Quay to avoid repetition...